src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -322,15 +359,35 @@ public void OnResponseAbort()
                 }
             }
 
-            public async Task ReadResponseHeadersAsync()
+            public async Task ReadResponseHeadersAsync(bool createStream, TaskCompletionSource<bool> tcs = null)
             {
                 // Wait for response headers to be read.
-                (bool wait, bool emptyResponse) = TryEnsureHeaders();
-                if (wait)
+                bool emptyResponse;
+                bool wait;
+
+                do
                 {
-                    await GetWaiterTask().ConfigureAwait(false);
                     (wait, emptyResponse) = TryEnsureHeaders();
-                    Debug.Assert(!wait);
+                    if (wait)
+                    {
+                        await GetWaiterTask().ConfigureAwait(false);
+                        (wait, emptyResponse) = TryEnsureHeaders();
+                        Debug.Assert(!wait);
+                    }
+
+                    if (Response.StatusCode  == HttpStatusCode.Continue)
+                    {
+                        OnResponse100Continue();
+                        tcs?.TrySetResult(true);
+                    }
+                }
+                while ((uint)Response.StatusCode < 200);
+
+                if (tcs != null && (int)Response.StatusCode >= 300)
```

Comment:
It seems like the tcs should be set to null when we successfully receive a 100 Continue response. Am I missing where this happens?

Reply:
I don't understand. We may not have received 100 or we may not do it at all. I simplified the section.